transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,370 @@
+// MIT License
+//
+// Copyright (c) 2021 WebRTC.rs
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use async_trait::async_trait;
+use futures::ready;
+use stun::{
+    attributes::ATTR_USERNAME,
+    message::{is_message as is_stun_message, Message as STUNMessage},
+};
+use tokio::{io::ReadBuf, net::UdpSocket};
+use tokio_crate as tokio;
+use webrtc_ice::udp_mux::{UDPMux, UDPMuxConn, UDPMuxConnParams, UDPMuxWriter};
+use webrtc_util::{sync::RwLock, Conn, Error};
+
+use std::{
+    collections::{HashMap, HashSet},
+    io::ErrorKind,
+    net::SocketAddr,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex, Weak,
+    },
+    task::{Context, Poll},
+};
+
+const RECEIVE_MTU: usize = 8192;
+
+/// A previously unseen address of a remote who've sent us an ICE binding request.
+#[derive(Debug)]
+pub struct NewAddr {
+    pub addr: SocketAddr,
+    pub ufrag: String,
+}
+
+/// An event emitted by [`UDPMuxNewAddr`] when it's polled.
+#[derive(Debug)]
+pub enum UDPMuxEvent {
+    /// Connection error. UDP mux should be stopped.
+    Error(std::io::Error),
+    /// Got a [`NewAddr`] from the socket.
+    NewAddr(NewAddr),
+}
+
+/// A modified version of [`webrtc_ice::udp_mux::UDPMuxDefault`], which reports previously unseen
+/// addresses instead of ignoring them.
+pub struct UDPMuxNewAddr {
+    udp_sock: UdpSocket,
+
+    /// Maps from ufrag to the underlying connection.
+    conns: Mutex<HashMap<String, UDPMuxConn>>,
+
+    /// Maps from socket address to the underlying connection.
+    address_map: RwLock<HashMap<SocketAddr, UDPMuxConn>>,
+
+    /// Set of the new addresses to avoid sending the same address multiple times.
+    new_addrs: RwLock<HashSet<SocketAddr>>,
+
+    /// `true` when UDP mux is closed.
+    is_closed: AtomicBool,
+}
```

Comment:
So I am wondering whether or not it is worth it to completely rewrite this thing into a single, stateful struct and a `Handle` with the `Handle` being the one implementing the `async_trait`s from `webrtc` and being connected to the struct via channels for the individual async fns. Essentially the way `rust-yamux` is designed.

I think that would remove these locks.

@mxinden Thoughts?

Reply:
I've made a dirty PoC for this here: https://github.com/melekes/rust-libp2p/pull/1

This is quite the substantial change to the copied code though. I think it is worth it and I am happy to rewrite the entire thing and make a PR to your fork @melekes.

What we need to decide on is whether we actually want this. IMO, vendoring code is fine but we need to treat it as our own and not optimise it for being able to replace it again when for example a new release of `webrtc-rs` comes out. The trade-off is that we then have to adapt to those changes ourselves.

Curious to hear your thoughts @melekes @mxinden @elenaf9.

Reply:
Thanks for the PoC 👍 While I can see the benefits here (no locks, better prioritization), can't this be viewed as a premature optimization? The code with locks works. Shouldn't we prove that there's a high lock contention before "rewriting the whole thing"?

> vendoring code is fine but we need to treat it as our own and not optimise it for being able to replace it again 

I don't think we'll ever need to replace it per se. It has already diverged from the original copy.

Reply:
> Shouldn't we prove that there's a high lock contention before \"rewriting the whole thing\"?

My motivation for rewriting it wasn't as much that the current code doesn't work but that it doesn't fit the style of the codebase.

Whether or not that reason is enough to justify the rewrite is what we need to agree on :)